src/index/base.cpp

```diff
@@ -230,18 +234,43 @@ bool BaseIndex::Rewind(const CBlockIndex* current_tip, const CBlockIndex* new_ti
 
 void BaseIndex::BlockConnected(const std::shared_ptr<const CBlock>& block, const CBlockIndex* pindex)
 {
+    return this->blockConnectedInternal(block, pindex, false);
+}
+
+void BaseIndex::BackgroundBlockConnected(
+    const std::shared_ptr<const CBlock>& block,
+    const CBlockIndex* pindex)
+{
+    this->blockConnectedInternal(block, pindex, true);
+}
+
+void BaseIndex::blockConnectedInternal(
+    const std::shared_ptr<const CBlock>& block,
+    const CBlockIndex* pindex,
+    bool from_background)
+{
+    // Ignore BlockConnected signals until we have fully indexed the chain.
     if (!m_synced) {
         return;
     }
 
+    bool bg_chainstates_in_use = m_chainman->hasBgChainstateInUse();
```

Comment:
The reason the indexing code needs to be aware of background chainstates is because the best block pointer (`m_best_block_index`) was (and is) used to signal that all blocks including and beneath that blockhash have been indexed. When you've got two separate chainstates throwing off BlockConnected events, you basically have to reserve updating that pointer for the background chainstate because otherwise you're breaking the nature of the best block pointer. Updating the index when you get an assumed-valid chainstate event would mean that there are unindexed blocks beneath the pointer.

I've tried to think about ways to avoid this situation, but haven't been able to figure out anything simpler. If you've got ideas, I'm definitely curious and happy to implement.

Reply:
Yeah I think what I'm asking is beyond the scope of the PR. It seems to me ideally validation code would notify indexing code "the tip is changing" and "here is some new block data" in 2 separate events not 3 mixed up events (block connected, background block connected, rewind), and also that indexing code would more passively receive data instead of querying chainstatemanager to reverse-engineer what validation code is doing.

But probably what you're doing now is the simplest approach for this PR.